pkg/adaptor/cloud/ibmcloud-powervs/provider.go

```diff
@@ -0,0 +1,228 @@
+	body := &models.PVMInstanceCreate{
+		ServerName:  &instanceName,
+		ImageID:     &p.serviceConfig.ImageID,
+		KeyPairName: p.serviceConfig.SSHKey,
+		Networks: []*models.PVMInstanceAddNetwork{
+			{
+				NetworkID: &p.serviceConfig.NetworkID,
+			}},
+		Memory:     core.Float64Ptr(p.serviceConfig.Memory),
+		Processors: core.Float64Ptr(p.serviceConfig.Processors),
+		ProcType:   core.StringPtr(p.serviceConfig.ProcessorType),
+		SysType:    p.serviceConfig.SystemType,
+		UserData:   base64.StdEncoding.EncodeToString([]byte(userData)),
+	}
+
+	logger.Printf("CreateInstance: name: %q", instanceName)
+
+	pvsInstances, err := p.powervsService.instanceClient.Create(body)
+	if err != nil {
+		logger.Printf("failed to create an instance : %v", err)
+		return nil, err
+	}
+
+	if len(*pvsInstances) <= 0 {
+		return nil, fmt.Errorf("there are no instances created")
+	}
+
+	ins := (*pvsInstances)[0]
+	instanceID := *ins.PvmInstanceID
+
+	num := 1
+	for {
+		in, err := p.powervsService.instanceClient.Get(*ins.PvmInstanceID)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get the instance: %v", err)
+		}
+
+		if *in.Status == "ERROR" {
+			return nil, fmt.Errorf("instance is in error state")
+		}
+
+		if *in.Status == "ACTIVE" {
+			logger.Printf("instance is in desired state: %s", *in.Status)
+			break
+		}
+		if num == maxRetries {
+			return nil, fmt.Errorf("reached maximum retry count for reaching desired instance state")
+		}
+		logger.Printf("[%d]Fetching current instance state: %s", num, *in.Status)
+		num++
+
+		time.Sleep(time.Duration(pollingInterval) * time.Second)
+	}
+
+	ips, err := p.getVMIPs(ins)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get IPs for the instance : %v", err)
+	}
+
+	return &cloud.Instance{
+		ID:   instanceID,
+		Name: instanceName,
+		IPs:  ips,
+	}, nil
+}
+
+func (p *ibmcloudPowerVSProvider) DeleteInstance(ctx context.Context, instanceID string) error {
+
+	err := p.powervsService.instanceClient.Delete(instanceID)
+	if err != nil {
+		logger.Printf("failed to delete an instance: %v", err)
+		return err
+	}
+
+	logger.Printf("deleted an instance %s", instanceID)
+	return nil
+}
+
+func (p *ibmcloudPowerVSProvider) Teardown() error {
+	return nil
+}
+
+func (p *ibmcloudPowerVSProvider) getVMIPs(instance *models.PVMInstance) ([]net.IP, error) {
+	var ips []net.IP
+	ins, err := p.powervsService.instanceClient.Get(*instance.PvmInstanceID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get the instance: %v", err)
+	}
+
+	for i, network := range ins.Networks {
+		if ins.Networks[i].Type == "fixed" {
+			ip := net.ParseIP(network.IPAddress)
+			if ip == nil {
+				return nil, fmt.Errorf("failed to parse pod node IP: %q", network.IPAddress)
+			}
+
+			ips = append(ips, ip)
+			logger.Printf("podNodeIP[%d]=%s", i, ip.String())
+		}
+	}
+
+	if len(ips) > 0 {
+		return ips, nil
+	}
+
+	// If IP is not assigned to the instance, fetch it from DHCP server
+	var ip *string
+	for i := 0; i < maxRetries; i++ {
```

Comment:
Same comment here. Can you use retry.Do with a timeout?

Reply:
Addressed it, PTAL.